### PR TITLE
 Expanded the CommonReadOnlySequenceTests 

### PR DESCRIPTION
**Pull request by AlexRadch**
I expanded the ReadOnlySequence Common Tests and some of them fails.
I think this is code issues.

### CODE REVIEW

**Comment by AlexRadch** [comment]
see https://github.com/dotnet/corefx/issues/26768

**Comment by ahsonkhan** [comment]
cc @pakrym, @davidfowl

File: src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs [hunk 1]
```diff
@@ -144,63 +156,111 @@ public void CanGetFirst()
...
         }
 
         [Fact]
         public void SeekEmptySkipDoesNotCrossPastEnd()
         {
-            var bufferSegment1 = new BufferSegment<byte>(new byte[100]);
-            BufferSegment<byte> bufferSegment2 = bufferSegment1.Append(new byte[0]);
-            BufferSegment<byte> bufferSegment3 = bufferSegment2.Append(new byte[0]);
-            BufferSegment<byte> bufferSegment4 = bufferSegment3.Append(new byte[100]);
+            var bufferSegment1 = new BufferSegment<byte>(new byte[0]);
```
Comment (davidfowl) [inline src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: Why did you change the existing test?

**Comment by davidfowl** [comment]
> I expanded the ReadOnlySequence Common Tests and some of them fails. I think this is code issues.

Can you be more specific?

File: src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs [hunk 2]
```diff
@@ -144,63 +156,111 @@ public void CanGetFirst()
...
+            Assert.Equal(bufferSegment8, c1.GetObject());
 
+            c1 = buffer.GetPosition(0, c1);
             Assert.Equal(0, c1.GetInteger());
-            Assert.Equal(bufferSegment2, c1.GetObject());
+            Assert.Equal(bufferSegment8, c1.GetObject());
 
-            c1 = buffer.GetPosition(100, buffer.Start);
+            c1 = buffer.GetPosition(250);
+            Assert.Equal(0, c1.GetInteger());
```
Comment (benaadams) [inline src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: This fails because 250 is 50 bytes into the second array with size; so it should be
```
Assert.Equal(50, c1.GetInteger());
```

File: src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs [hunk 3]
```diff
@@ -144,63 +156,111 @@ public void CanGetFirst()
...
+            Assert.Equal(bufferSegment8, c1.GetObject());
 
+            c1 = buffer.GetPosition(250, buffer.Start);
             Assert.Equal(0, c1.GetInteger());
```
Comment (benaadams) [inline src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: Also this would need to be
```
Assert.Equal(50, c1.GetInteger());
```

File: src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs [hunk 4]
```diff
@@ -144,63 +156,111 @@ public void CanGetFirst()
...
-            Assert.Equal(bufferSegment2, c1.GetObject());
+            Assert.Equal(bufferSegment8, c1.GetObject());
+
+            c1 = buffer.GetPosition(50, c1);
+            Assert.Equal(0, c1.GetInteger());
```
Comment (benaadams) [inline src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: As would this
```
Assert.Equal(50, c1.GetInteger());
```

**Comment by pakrym** [comment]
Let's not expand existing tests, if there is some coverage missing let's add another test that fixes that tests some concrete scenario.

File: src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Long.cs [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+        private static long GetRunningIndex<T>(SequencePosition position) => ((BufferSegment<T>)position.GetObject()).RunningIndex;
+        private static int GetSegmentsCount<T>(ReadOnlySequence<T> buffer) =>
+            (int)((buffer.Length + buffer.Start.GetInteger() + SegmentLength - 1) / SegmentLength);
+
+        private static long GetOffset(int segment, int integer, int startOffset = DefaultStartOffset) => (long)segment * SegmentLength + integer - startOffset;
+
+        private readonly ReadOnlySequence<byte> _byteSequence = NewSequence<byte>();
+
+        [Fact]
+        public void SegmentStartIsConsideredInBoundsCheck()
```
Comment (pakrym) [inline src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Long.cs]: Why are only 3 of these tests run on the large sequence?

File: src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs [hunk 6]
```diff
@@ -80,27 +80,39 @@ public void GetPositionDoesNotCrossOutsideBuffer()
...
             c1 = buffer.GetPosition(200, buffer.Start);
-
             Assert.Equal(0, c1.GetInteger());
             Assert.Equal(bufferSegment3, c1.GetObject());
 
             seq = buffer.Slice(0, c1);
             Assert.Equal(200, seq.Length);
+
+            c1 = buffer.GetPosition(350);
+            Assert.Equal(100, c1.GetInteger());
```
Comment (benaadams) [inline src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: 350 is 50 into the 3rd 100 array so it would be 50 rather than 100

File: src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs (hunk 2 above)
Comment (AlexRadch) [reply src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: @benaadams `c1 = buffer.GetPosition(250);` should 
1. return position behind End
2. stop on End and return End
3. Throw some Exception?

What is correct?

Comment (benaadams) [reply src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: Ah, see what you are doing...

This might be incorrect behaviour by `ReadOnlySequence`

Comment (benaadams) [reply src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: It should throw `ArgumentOutOfRangeException` (and does I believe?)

Comment (benaadams) [reply src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: Ah.. looks like there is a corner case not handled

Comment (benaadams) [reply src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: Added fix https://github.com/dotnet/corefx/pull/28261

Comment (ahsonkhan) [reply src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: I borrowed the tests from https://github.com/dotnet/corefx/pull/28261 that catch this issue in the following PR: https://github.com/dotnet/corefx/pull/28294
